python/cuml/random_forest/random_forest.pyx

```diff
@@ -0,0 +1,213 @@
+        
+
+
+class Randomforest(Base):
+    """
+    Description and example code
+
+    """
+    
+    # min_rows_per_node in cuml = min_samples_split in sklearn
+    # max_leaves
+    def __init__(self, n_estimators=25, max_depth=None, max_features=None, min_rows_per_node=None, bootstrap=True):
+        self.n_estimators = n_estimators
+        self.max_depth = max_depth
+        self.max_features = max_features
+        self.min_rows_per_node = min_rows_per_node
+        self.bootstrap = bootstrap
+
+    def _get_ctype_ptr(self, obj):
+        # The manner to access the pointers in the gdf's might change, so
+        # encapsulating access in the following 3 methods. They might also be
+        # part of future gdf versions.
+        return obj.device_ctypes_pointer.value
+
+    def _get_column_ptr(self, obj):
+        return self._get_ctype_ptr(obj._column._data.to_gpu_array())
+
+    def fit(self, X):
+
+        if self.labels_ is not None:
+            del self.labels_
+
+        if self.labels_array is not None:
+            del self.labels_array
+
+        cdef uintptr_t input_ptr
+        if (isinstance(X, cudf.DataFrame)):
+            self.gdf_datatype = np.dtype(X[X.columns[0]]._column.dtype)
+            X_m = numba_utils.row_matrix(X)
+            self.n_rows = len(X)
+            self.n_cols = len(X._cols)
+
+        elif (isinstance(X, np.ndarray)):
+            self.gdf_datatype = X.dtype
+            X_m = cuda.to_device(X)
+            self.n_rows = X.shape[0]
+            self.n_cols = X.shape[1]
+
+
+        else:
+            msg = "X matrix format  not supported"
+            raise TypeError(msg)
+
+        input_ptr = self._get_ctype_ptr(X_m)
+
+        cdef cumlHandle* handle_ = <cumlHandle*> <size_t> self.handle.getHandle()
+        self.labels_ = cudf.Series(np.zeros(self.n_rows, dtype=np.int32))
+        self.labels_array = self.labels_._column._data.to_gpu_array()
+        cdef uintptr_t labels_ptr = self._get_ctype_ptr(self.labels_array)
+
+        cdef rfClassifier rf_classifier
+
+        if self.gdf_datatype.type == np.float32:
+            fit(handle_[0],
+                <float*> rf_classifier,
+                <float*> input_ptr,
+                <int> self.n_rows,
+                <int> self.n_cols,
+                <int*> labels_ptr)
+        else:
+            fit(handle_[0],
+                <float*> rf_classifier,
+                <float*> input_ptr,
+                <int> self.n_rows,
+                <int> self.n_cols,
+                <int*> labels_ptr)
+
+        # make sure that the `fit` is complete before the following delete
+        # call happens
+        self.handle.sync()
+        del(X_m)
+        return self
+
+
+    # n_unique_labels is found while running the preprocess_labels function 
+
+    def predict(self, X):
+       
+        cdef uintptr_t input_ptr
+        if (isinstance(X, cudf.DataFrame)):
+            self.gdf_datatype = np.dtype(X[X.columns[0]]._column.dtype)
```

Comment:
Not sure that predict should set anything on self. I think it's surprising if predict changes any internal state since you'll often generate one instance and call predict many times on it.

Reply:
So it should probably be more like checking that the dtype here matches the expected self.dtype

Reply:
Having a self.dtype (instead of the legacy badly named `gdf_datatype`) for all models to be able to check inputs is being standardized on PR #612 

This PR can either follow the example there if that one is merged first, or I can change this in that PR if this one makes it first

Reply:
I can edit it to follow the PR #612